### PR TITLE
feat: compute rollup gas price for StateTransition

### PR DESCRIPTION
**Pull request by 0xmountaintop**
reference:
+ https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/core/state_transition.go#L133
+ https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/rollup/fees/rollup_fee.go#L121
+ ~~https://github.com/scroll-tech/scroll/blob/feat/deposit_and_withdraw_proof/contracts/src/L2/predeploys/IL2GasPriceOracle.sol~~
+ ~~https://github.com/scroll-tech/scroll/blob/feat/deposit_and_withdraw_proof/contracts/src/L2/predeploys/L1BlockContainer.sol~~
+ https://github.com/scroll-tech/scroll/blob/feat/new_bridge_contracts/contracts/src/L2/predeploys/L1GasPriceOracle.sol
+ https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts/contracts/L2/predeploys/OVM_GasPriceOracle.sol

### CODE REVIEW

File: rollup/rcfg/config.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package rcfg
+
+import (
+	"math/big"
+
+	"github.com/scroll-tech/go-ethereum/common"
+)
+
+// UsingSVM is used to enable or disable functionality necessary for the SVM.
+var UsingSVM bool
```
Comment (icemelon) [inline rollup/rcfg/config.go]: Prefer to use some other name like `UsingScroll`. `SVM` gives others the impression that we are not compatible with `EVM`.

**Comment by zimpha** [comment]
l1BaseFee slot = 1
overhead slot = 2
scalar slot = 3

@HAOYUatHZ

File: rollup/fees/rollup_fee.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// CalculateL1GasUsed computes the L1 gas used based on the calldata and
+// constant sized overhead. The overhead can be decreased as the cost of the
+// batch submission goes down via contract optimizations. This will not overflow
+// under standard network conditions.
+func CalculateL1GasUsed(data []byte, overhead *big.Int) *big.Int {
+	zeroes, ones := zeroesAndOnes(data)
+	zeroesGas := zeroes * params.TxDataZeroGas
+	onesGas := (ones + 68) * params.TxDataNonZeroGasEIP2028
```
Comment (Thegaram) [inline rollup/fees/rollup_fee.go:126]: What does this `+ 68` represent?

**Review by Thegaram — DISMISSED** [review]

File: rollup/fees/rollup_fee.go (hunk 2 above)
Comment (icemelon) [reply rollup/fees/rollup_fee.go:126]: tx header size

File: core/state_transition.go [hunk 3]
```diff
@@ -336,7 +352,17 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
...
 		effectiveTip = cmath.BigMin(st.gasTipCap, new(big.Int).Sub(st.gasFeeCap, st.evm.Context.BaseFee))
 	}
-	st.state.AddBalance(st.evm.FeeRecipient(), new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), effectiveTip))
+
+	if st.evm.ChainConfig().UsingScroll {
+		// The L2 Fee is the same as the fee that is charged in the normal geth
+		// codepath. Add the L1 fee to the L2 fee for the total fee that is sent
+		// to the sequencer.
+		l2Fee := new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), effectiveTip)
+		fee := new(big.Int).Add(st.l1Fee, l2Fee)
```
Comment (mask-pp) [inline core/state_transition.go:363]:
```suggestion
		effectiveTip.Mul(effectiveTip, new(big.Int).SetUint64(st.gasUsed()))
```

Comment (mask-pp) [inline core/state_transition.go:363]:
```suggestion
		effectiveTip.Add(effectiveTip, st.l1Fee)
```

File: core/state_transition.go [hunk 4]
```diff
@@ -336,7 +352,17 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
...
+		st.state.AddBalance(st.evm.FeeRecipient(), fee)
+	} else {
+		st.state.AddBalance(st.evm.FeeRecipient(), new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), effectiveTip))
```
Comment (mask-pp) [inline core/state_transition.go:364]: Delete line362, and don't need to use `else` logic.

File: rollup/fees/rollup_fee.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	// errTransactionSigned represents the error case of passing in a signed
+	// transaction to the L1 fee calculation routine. The signature is accounted
+	// for externally
+	errTransactionSigned = errors.New("transaction is signed")
+)
+
+// Message represents the interface of a message.
+// It should be a subset of the methods found on
+// types.Message
+type Message interface {
```
Comment (mask-pp) [inline rollup/fees/rollup_fee.go:25]: Can reuse `core.Message` interface.

**Review by mask-pp — DISMISSED** [review]
LGTM

File: rollup/fees/rollup_fee.go (hunk 2 above)
Comment (Thegaram) [reply rollup/fees/rollup_fee.go:126]: I see, it's for the signature. Explained here: https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts/contracts/L2/predeploys/OVM_GasPriceOracle.sol#L139-L145

**Review by mask-pp — APPROVED** [review]

**Pull request merged**